### DISCUSSION TITLE
How to run steps involving environment variables on both Linux and Windows

### DISCUSSION THREAD

**Discussion by jonhue** [community/community#25071] [Actions]
<pre><code>build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os:
          - ubuntu-latest
          - windows-latest

    steps:
    - uses: actions/checkout@master
    - uses: actions/setup-node@v1
      with:
        node-version: 12
    - name: Install dependencies
      run: yarn install --dev

    - name: Build binaries and upload to GitHub
      if: shell == 'bash'
      run: yarn prebuild --all -u $PREBUILD_TOKEN
      env:
        PREBUILD_TOKEN: ${{ secrets.PREBUILD_TOKEN }}
    - name: Build binaries and upload to GitHub
      if: shell == 'powershell'
      run: yarn prebuild --all -u ${env:PREBUILD_TOKEN}
      env:
        PREBUILD_TOKEN: ${{ secrets.PREBUILD_TOKEN }}
</code></pre>
<p>I have this job that is supposed to upload builds to GitHub releases. It runs on both Linux and Windows. As Powershell is used as a shell for Windows builds, I have to use a different syntax to read the environment variable that stores credentials to upload to GitHub. Otherwise, my Windows builds would just fail. So I tried to add conditional steps. But now GitHub doesn’t run the action at all and instead errors with:</p>
<pre><code>### ERRORED 13:33:33Z

- Your workflow file was invalid: The pipeline is not valid. .github/workflows/prebuild.yml (Line: 30, Col: 11): Unrecognized named-value: 'shell'. Located at position 1 within expression: shell == 'bash',.github/workflows/prebuild.yml (Line: 35, Col: 11): Unrecognized named-value: 'shell'. Located at position 1 within expression: shell == 'powershell'
</code></pre>
<p>What is the suggested way of running the same steps in different shells when access to environment variables is necessary? </p>

**Comment by weide-zhou** [comment, marked as answer]
<p>‘shell’ is not recognized because it’s not the value of workflow context. As a workaround, you can use below condition instead.</p>
<p>Typically you can use the context value for the condition filter when access to environment variable is necessary. Please check the link <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#contexts" rel="nofollow noopener">here</a> for more details.</p>
<pre><code>if: runner.os == 'Linux'
… bash command
If: runner.os == 'Windows'
… powershelll command
</code></pre>
<p>or </p>
<pre><code>if: matrix.os == 'ubuntu-latest'
… bash command
If: matrix.os == 'windows-latest'
… powershelll command
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/9/915d91435e7811e66c059dc4c79bbc44ad99bd83.png" data-download-href="/uploads/short-url/kJXHqN905BPJxnOrJdZzifxn5wT.png?dl=1" title="context.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181088654-797a699e-0e0d-4209-98f2-8da17e6c3b4d.png" alt="context.png" data-base62-sha1="kJXHqN905BPJxnOrJdZzifxn5wT" width="690" height="457" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/915d91435e7811e66c059dc4c79bbc44ad99bd83_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">context.png</span><span class="informations">733×486 20.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
